In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/human-activity-recognition-with-smartphones/train.csv
/kaggle/input/human-activity-recognition-with-smartphones/test.csv


In [37]:
df = pd.read_csv('/kaggle/input/human-activity-recognition-with-smartphones/train.csv').drop(columns='subject')

In [4]:
df.head()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0           0.288585          -0.020294          -0.132905         -0.995279   
1           0.278419          -0.016411          -0.123520         -0.998245   
2           0.279653          -0.019467          -0.113462         -0.995380   
3           0.279174          -0.026201          -0.123283         -0.996091   
4           0.276629          -0.016570          -0.115362         -0.998139   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
0         -0.983111         -0.913526         -0.995112         -0.983185   
1         -0.975300         -0.960322         -0.998807         -0.974914   
2         -0.967187         -0.978944         -0.996520         -0.963668   
3         -0.983403         -0.990675         -0.997099         -0.982750   
4         -0.980817         -0.990482         -0.998321         -0.979672   

   tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-skewness()  \
0         -0.923527         -0.934724  ...                        -0.298676   
1         -0.957686         -0.943068  ...                        -0.595051   
2         -0.977469         -0.938692  ...                        -0.390748   
3         -0.989302         -0.938692  ...                        -0.117290   
4         -0.990441         -0.942469  ...                        -0.351471   

   fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
0                        -0.710304                    -0.112754   
1                        -0.861499                     0.053477   
2                        -0.760104                    -0.118559   
3                        -0.482845                    -0.036788   
4                        -0.699205                     0.123320   

   angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
0                              0.030400                         -0.464761   
1                             -0.007435                         -0.732626   
2                              0.177899                          0.100699   
3                             -0.012892                          0.640011   
4                              0.122542                          0.693578   

   angle(tBodyGyroJerkMean,gravityMean)  angle(X,gravityMean)  \
0                             -0.018446             -0.841247   
1                              0.703511             -0.844788   
2                              0.808529             -0.848933   
3                             -0.485366             -0.848649   
4                             -0.615971             -0.847865   

   angle(Y,gravityMean)  angle(Z,gravityMean)  Activity  
0              0.179941             -0.058627  STANDING  
1              0.180289             -0.054317  STANDING  
2              0.180637             -0.049118  STANDING  
3              0.181935             -0.047663  STANDING  
4              0.185151             -0.043892  STANDING  

[5 rows x 562 columns]

In [6]:
df.shape

(7352, 562)

In [38]:
df['Activity'].value_counts()

Activity
LAYING                1407
STANDING              1374
SITTING               1286
WALKING               1226
WALKING_UPSTAIRS      1073
WALKING_DOWNSTAIRS     986
Name: count, dtype: int64

In [7]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import LabelEncoder

In [39]:
x = df.drop(columns=['Activity'])
y = df['Activity']

le = LabelEncoder()
y = le.fit_transform(y)

x_train,x_test, y_train, y_test = train_test_split(x,y,test_size=0.2,random_state=42)

In [40]:
lg = LogisticRegression(max_iter=1000)

In [41]:
lg.fit(x_train,y_train)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=1000)

### Droping duplicate columns


In [42]:
y_pred = lg.predict(x_test)

In [43]:
print("Accuracy metrics : ",accuracy_score(y_test,y_pred))

Accuracy metrics :  0.9809653297076818


### Now we remove reduce feature 562 to 100
### And then see is how much accuracy affected

In [44]:
# Removing Duplicates

def get_duplicate_columns(df):
    
    duplicate_columns = {}
    seen_columns = {}

    for column in df.columns:
        current_column = df[column]

        # Convert column data to bytes
        try:
            current_column_hash = current_column.values.tobytes()
        except AttributeError:
            current_column_hash = current_column.to_string().encode()

        if current_column_hash in seen_columns:
            if seen_columns[current_column_hash] in duplicate_columns:
                duplicate_columns[seen_columns[current_column_hash]].append(column)
            else:
                duplicate_columns[seen_columns[current_column_hash]] = [column]
        else:
            seen_columns[current_column_hash] = column

    return duplicate_columns

In [51]:
duplicate_columns = get_duplicate_columns(x_train)
duplicate_columns

{}

In [52]:
# now we just keep key feature and remove all duplicate which in values

for i in duplicate_columns.values():
    x_train.drop(columns=i, inplace=True)
    x_test.drop(columns=i, inplace=True)

In [48]:
x_train.head()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
57             0.278007          -0.017803          -0.108965   
4154           0.237617          -0.000782          -0.114476   
6945           0.290924          -0.050878          -0.073518   
527            0.275268          -0.015050          -0.114204   
4196           0.278790          -0.018585          -0.106908   

      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
57           -0.994425         -0.994873         -0.994886         -0.994939   
4154         -0.326331          0.069663         -0.224321         -0.343326   
6945         -0.026220         -0.032163          0.393109         -0.118256   
527          -0.981092         -0.901124         -0.960423         -0.984417   
4196         -0.997380         -0.983893         -0.984482         -0.997331   

      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
57           -0.993994         -0.995450         -0.938974  ...   
4154          0.039623         -0.256327         -0.310961  ...   
6945         -0.030279          0.432861          0.370607  ...   
527          -0.901405         -0.965788         -0.922291  ...   
4196         -0.985196         -0.983768         -0.942062  ...   

      fBodyBodyGyroJerkMag-meanFreq()  fBodyBodyGyroJerkMag-skewness()  \
57                           0.394506                        -0.684497   
4154                         0.052089                        -0.131946   
6945                        -0.038923                        -0.643660   
527                         -0.145084                        -0.326860   
4196                         0.096524                        -0.113724   

      fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
57                          -0.901888                     0.751089   
4154                        -0.591721                     0.628099   
6945                        -0.935668                    -0.182993   
527                         -0.645937                     0.205896   
4196                        -0.445144                    -0.267371   

      angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
57                                0.626432                          0.263401   
4154                             -0.357089                         -0.925512   
6945                              0.484857                         -0.961034   
527                              -0.168239                         -0.197557   
4196                             -0.304399                         -0.081555   

      angle(tBodyGyroJerkMean,gravityMean)  angle(X,gravityMean)  \
57                               -0.646225              0.398881   
4154                              0.110212             -0.474013   
6945                              0.147325             -0.819001   
527                               0.328892             -0.744428   
4196                              0.110146             -0.615329   

      angle(Y,gravityMean)  angle(Z,gravityMean)  
57               -0.546717             -0.431200  
4154              0.383162              0.238969  
6945              0.221549             -0.003571  
527               0.274256              0.031246  
4196              0.322236              0.161577  

[5 rows x 540 columns]

In [53]:
x_train.shape

(5881, 540)

### 1. Variance Threshold

In [54]:
from sklearn.feature_selection import VarianceThreshold
vt = VarianceThreshold(threshold=0.05)

In [55]:
# ab yha par model hamare x_train me har column ka varianc calculate krega
# jinka bhi variance < 0.05 ----> unko drop kar dega

vt.fit(x_train)

VarianceThreshold(threshold=0.05)

In [56]:
vt.get_support().sum()

# it tell ki kitne columns threshold se upper hai 
# total = 349  ----> only keep these columns

349

In [58]:
columns = x_train.columns[vt.get_support()]
columns

Index(['tBodyAcc-std()-X', 'tBodyAcc-std()-Y', 'tBodyAcc-std()-Z',
       'tBodyAcc-mad()-X', 'tBodyAcc-mad()-Y', 'tBodyAcc-mad()-Z',
       'tBodyAcc-max()-X', 'tBodyAcc-max()-Y', 'tBodyAcc-max()-Z',
       'tBodyAcc-min()-X',
       ...
       'fBodyBodyGyroJerkMag-meanFreq()', 'fBodyBodyGyroJerkMag-skewness()',
       'fBodyBodyGyroJerkMag-kurtosis()', 'angle(tBodyAccMean,gravity)',
       'angle(tBodyAccJerkMean),gravityMean)',
       'angle(tBodyGyroMean,gravityMean)',
       'angle(tBodyGyroJerkMean,gravityMean)', 'angle(X,gravityMean)',
       'angle(Y,gravityMean)', 'angle(Z,gravityMean)'],
      dtype='object', length=349)

In [61]:
x_train = vt.transform(x_train)
x_test = vt.transform(x_test)

x_train = pd.DataFrame(x_train, columns=columns)
x_test = pd.DataFrame(x_test, columns=columns)

### Covariance

In [62]:
x_train.shape

(5881, 349)

In [63]:
corr_matrix = x_train.corr()

In [64]:
corr_matrix

tBodyAcc-std()-X  tBodyAcc-std()-Y  \
tBodyAcc-std()-X                              1.000000          0.927247   
tBodyAcc-std()-Y                              0.927247          1.000000   
tBodyAcc-std()-Z                              0.850268          0.895065   
tBodyAcc-mad()-X                              0.998631          0.922627   
tBodyAcc-mad()-Y                              0.920936          0.997384   
...                                                ...               ...   
angle(tBodyGyroMean,gravityMean)              0.023914         -0.002241   
angle(tBodyGyroJerkMean,gravityMean)         -0.035176         -0.028881   
angle(X,gravityMean)                         -0.374114         -0.383095   
angle(Y,gravityMean)                          0.472605          0.524945   
angle(Z,gravityMean)                          0.393209          0.432180   

                                      tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
tBodyAcc-std()-X                              0.850268          0.998631   
tBodyAcc-std()-Y                              0.895065          0.922627   
tBodyAcc-std()-Z                              1.000000          0.842986   
tBodyAcc-mad()-X                              0.842986          1.000000   
tBodyAcc-mad()-Y                              0.890973          0.916201   
...                                                ...               ...   
angle(tBodyGyroMean,gravityMean)             -0.010535          0.024098   
angle(tBodyGyroJerkMean,gravityMean)         -0.016002         -0.035629   
angle(X,gravityMean)                         -0.344114         -0.370629   
angle(Y,gravityMean)                          0.475241          0.467965   
angle(Z,gravityMean)                          0.480824          0.389139   

                                      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  \
tBodyAcc-std()-X                              0.920936          0.845200   
tBodyAcc-std()-Y                              0.997384          0.894128   
tBodyAcc-std()-Z                              0.890973          0.997414   
tBodyAcc-mad()-X                              0.916201          0.838010   
tBodyAcc-mad()-Y                              1.000000          0.890707   
...                                                ...               ...   
angle(tBodyGyroMean,gravityMean)             -0.005865         -0.014838   
angle(tBodyGyroJerkMean,gravityMean)         -0.026679         -0.016949   
angle(X,gravityMean)                         -0.379578         -0.346350   
angle(Y,gravityMean)                          0.526803          0.476498   
angle(Z,gravityMean)                          0.430548          0.477627   

                                      tBodyAcc-max()-X  tBodyAcc-max()-Y  \
tBodyAcc-std()-X                              0.981284          0.893743   
tBodyAcc-std()-Y                              0.917831          0.953852   
tBodyAcc-std()-Z                              0.852711          0.864716   
tBodyAcc-mad()-X                              0.973704          0.888702   
tBodyAcc-mad()-Y                              0.911283          0.950131   
...                                                ...               ...   
angle(tBodyGyroMean,gravityMean)              0.029230         -0.000207   
angle(tBodyGyroJerkMean,gravityMean)         -0.038935         -0.013144   
angle(X,gravityMean)                         -0.386159         -0.373556   
angle(Y,gravityMean)                          0.482312          0.489971   
angle(Z,gravityMean)                          0.404088          0.424181   

                                      tBodyAcc-max()-Z  tBodyAcc-min()-X  ...  \
tBodyAcc-std()-X                              0.843918         -0.966714  ...   
tBodyAcc-std()-Y                              0.882782         -0.937472  ...   
tBodyAcc-std()-Z                              0.936311         -0.861033  ...   
tBodyAcc-mad()-X                              0.838024        

In [65]:
corr_matrix.columns

Index(['tBodyAcc-std()-X', 'tBodyAcc-std()-Y', 'tBodyAcc-std()-Z',
       'tBodyAcc-mad()-X', 'tBodyAcc-mad()-Y', 'tBodyAcc-mad()-Z',
       'tBodyAcc-max()-X', 'tBodyAcc-max()-Y', 'tBodyAcc-max()-Z',
       'tBodyAcc-min()-X',
       ...
       'fBodyBodyGyroJerkMag-meanFreq()', 'fBodyBodyGyroJerkMag-skewness()',
       'fBodyBodyGyroJerkMag-kurtosis()', 'angle(tBodyAccMean,gravity)',
       'angle(tBodyAccJerkMean),gravityMean)',
       'angle(tBodyGyroMean,gravityMean)',
       'angle(tBodyGyroJerkMean,gravityMean)', 'angle(X,gravityMean)',
       'angle(Y,gravityMean)', 'angle(Z,gravityMean)'],
      dtype='object', length=349)

In [66]:
columns = corr_matrix.columns

columns_to_drop = []

for i in range(len(columns)):
    for j in range(i + 1, len(columns)):
        if corr_matrix.loc[columns[i], columns[j]] > 0.90:
            columns_to_drop.append(columns[j])

print(len(columns_to_drop))

5583


In [67]:
columns_to_drop = set(columns_to_drop)

In [68]:
print(len(columns_to_drop))

221


In [69]:
x_train.drop(columns= columns_to_drop, axis=1, inplace=True)
x_test.drop(columns= columns_to_drop, axis=1, inplace=True)

In [70]:
x_train.shape

(5881, 128)

### ANOVA

In [74]:
from sklearn.feature_selection import f_classif
from sklearn.feature_selection import SelectKBest

# it give 100 best columns name jo hame rkhne chahiye
sel = SelectKBest(f_classif, k=100).fit(x_train, y_train)

# display selected feature names
x_train.columns[sel.get_support()]

Index(['tBodyAcc-std()-X', 'tBodyAcc-std()-Z', 'tBodyAcc-min()-X',
       'tBodyAcc-min()-Z', 'tBodyAcc-entropy()-X', 'tBodyAcc-entropy()-Y',
       'tBodyAcc-entropy()-Z', 'tBodyAcc-arCoeff()-X,1',
       'tBodyAcc-arCoeff()-X,2', 'tBodyAcc-arCoeff()-X,3',
       'tBodyAcc-arCoeff()-X,4', 'tBodyAcc-arCoeff()-Y,1',
       'tBodyAcc-arCoeff()-Z,1', 'tBodyAcc-arCoeff()-Z,3',
       'tBodyAcc-arCoeff()-Z,4', 'tBodyAcc-correlation()-X,Y',
       'tBodyAcc-correlation()-Y,Z', 'tGravityAcc-mean()-X',
       'tGravityAcc-mean()-Y', 'tGravityAcc-mean()-Z', 'tGravityAcc-sma()',
       'tGravityAcc-energy()-Y', 'tGravityAcc-energy()-Z',
       'tGravityAcc-entropy()-X', 'tGravityAcc-entropy()-Y',
       'tGravityAcc-arCoeff()-Y,1', 'tGravityAcc-arCoeff()-Y,2',
       'tGravityAcc-arCoeff()-Z,1', 'tGravityAcc-arCoeff()-Z,2',
       'tGravityAcc-correlation()-X,Y', 'tGravityAcc-correlation()-X,Z',
       'tGravityAcc-correlation()-Y,Z', 'tBodyAccJerk-arCoeff()-X,3',
       'tBodyAccJerk-arCoeff()-

In [76]:
columns = x_train.columns[sel.get_support()]


In [78]:
columns.shape

(100,)

In [79]:
# now keeping only those 100 columns

x_train = sel.transform(x_train)
x_test = sel.transform(x_test)

x_train = pd.DataFrame(x_train, columns=columns)
x_test = pd.DataFrame(x_test, columns=columns)

In [80]:
x_train.shape

(5881, 100)

In [82]:
log_reg = LogisticRegression(max_iter=1000)  # Increase max_iter if it doesn't converge
log_reg.fit(x_train, y_train)

# Make predictions on the test set
y_pred = log_reg.predict(x_test)

# Calculate and print accuracy score
accuracy = accuracy_score(y_test, y_pred)
print("Test accuracy:", accuracy)

Test accuracy: 0.9653297076818491


### Chi square test